src/sparse_blas/backends/mkl_common/mkl_handles.cxx

```diff
@@ -0,0 +1,396 @@
+                                      internal_smhandle->row_container.get_buffer<intType>(),
+                                      internal_smhandle->col_container.get_buffer<intType>(),
+                                      internal_smhandle->value_container.get_buffer<fpType>());
+    *p_smhandle = reinterpret_cast<oneapi::mkl::sparse::matrix_handle_t>(internal_smhandle);
+}
+
+template <typename fpType, typename intType>
+void init_csr_matrix(sycl::queue &queue, oneapi::mkl::sparse::matrix_handle_t *p_smhandle,
+                     std::int64_t num_rows, std::int64_t num_cols, std::int64_t /*nnz*/,
+                     oneapi::mkl::index_base index, intType *row_ptr, intType *col_ind,
+                     fpType *val) {
+    oneapi::mkl::sparse::matrix_handle_t mkl_handle;
+    oneapi::mkl::sparse::init_matrix_handle(&mkl_handle);
+    auto internal_smhandle = new detail::sparse_matrix_handle(mkl_handle, row_ptr, col_ind, val);
+    // The backend deduces nnz from row_ptr.
+    auto event = oneapi::mkl::sparse::set_csr_data(
+        queue, mkl_handle, static_cast<intType>(num_rows), static_cast<intType>(num_cols), index,
+        row_ptr, col_ind, val);
+    event.wait_and_throw();
+    *p_smhandle = reinterpret_cast<oneapi::mkl::sparse::matrix_handle_t>(internal_smhandle);
+}
+
+template <typename fpType, typename intType>
+void set_csr_matrix_data(sycl::queue &queue, oneapi::mkl::sparse::matrix_handle_t smhandle,
+                         std::int64_t num_rows, std::int64_t num_cols, std::int64_t /*nnz*/,
+                         oneapi::mkl::index_base index, sycl::buffer<intType, 1> row_ptr,
+                         sycl::buffer<intType, 1> col_ind, sycl::buffer<fpType, 1> val) {
+    auto internal_smhandle = detail::get_internal_handle(smhandle);
+    check_can_reset_sparse_handle<fpType, intType>(__FUNCTION__, internal_smhandle);
+    internal_smhandle->row_container.set_buffer(row_ptr);
+    internal_smhandle->col_container.set_buffer(col_ind);
+    internal_smhandle->value_container.set_buffer(val);
+    // The backend deduces nnz from row_ptr.
+    // The backend handle must use the buffers from the internal handle as they will be kept alive until the handle is released.
+    oneapi::mkl::sparse::set_csr_data(queue, internal_smhandle->backend_handle,
+                                      static_cast<intType>(num_rows),
+                                      static_cast<intType>(num_cols), index,
+                                      internal_smhandle->row_container.get_buffer<intType>(),
+                                      internal_smhandle->col_container.get_buffer<intType>(),
+                                      internal_smhandle->value_container.get_buffer<fpType>());
+}
+
+template <typename fpType, typename intType>
+void set_csr_matrix_data(sycl::queue &queue, oneapi::mkl::sparse::matrix_handle_t smhandle,
+                         std::int64_t num_rows, std::int64_t num_cols, std::int64_t /*nnz*/,
```

Comment:
if nnz is getting deduced, do we need std::int64_t arg declaration for nnz ?

Reply:
Yes, this is needed for cuSPARSE and rocSPARSE backends. They require `nnz` to be on the host. `nnz` can be deduced from the last value of `row_ptr` but this argument lets us avoid the cost of launching a copy from device to host.